aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use std::cmp;
+use std::ops::DerefMut;
+use std::sync::Mutex;
+
+use bytes::{Buf, Bytes, BytesMut};
+
+use crate::io::error::Error;
+use crate::io::path_body::PathBody;
+use crate::io::stream::RawInputStream;
+use crate::io::InputStream;
+use crate::MEBIBYTE;
+
+/// Builder for creating a `ReadPart` implementation.
+#[derive(Debug)]
+pub(crate) struct PartReaderBuilder {
+    stream: Option<RawInputStream>,
+    part_size: usize,
+}
+
+impl PartReaderBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            stream: None,
+            part_size: 5 * MEBIBYTE as usize,
+        }
+    }
+
+    /// Set the input stream to read from.
+    pub(crate) fn stream(mut self, stream: InputStream) -> Self {
+        self.stream = Some(stream.inner);
+        self
+    }
+
+    /// Set the target part size that should be used when reading data.
+    ///
+    /// All parts except for possibly the last one should be of this size.
+    pub(crate) fn part_size(mut self, part_size: usize) -> Self {
+        self.part_size = part_size;
+        self
+    }
+
+    pub(crate) fn build(self) -> impl ReadPart {
+        let stream = self.stream.expect("input stream set");
+        match stream {
+            RawInputStream::Buf(buf) => {
+                PartReader::Bytes(BytesPartReader::new(buf, self.part_size))
+            }
+            RawInputStream::Fs(path_body) => {
+                PartReader::Fs(PathBodyPartReader::new(path_body, self.part_size))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum PartReader {
+    Bytes(BytesPartReader),
+    Fs(PathBodyPartReader),
+}
+
+impl ReadPart for PartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        match self {
+            PartReader::Bytes(bytes) => bytes.next_part().await,
+            PartReader::Fs(path_body) => path_body.next_part().await,
+        }
+    }
+}
+
+/// Data for a single part
+pub(crate) struct PartData {
+    pub(crate) data: Bytes,
+    pub(crate) part_number: u64,
+}
```

Comment:
I would simplify this as:
```suggestion
/// Data for a single part
pub(crate) struct Part {
    pub(crate) data: Bytes,
    pub(crate) number: u64,
}
```
but maybe that's too far? One other thing to consider with this struct is the [memory layout](https://doc.rust-lang.org/nomicon/repr-rust.html). I think the `part_number` field should be first, otherwise there may end up being 32 bits of padding between `data` and `part_number` (although probably not b/c everything is 64bit now.)